apps/server/lib/lexical/server/code_intelligence/completion/translations/module_or_behaviour.ex

```diff
@@ -20,40 +20,50 @@ defmodule Lexical.Server.CodeIntelligence.Completion.Translations.ModuleOrBehavi
   end
 
   defp do_translate(%_{} = module, builder, %Env{} = env) do
-    struct_reference? = Env.in_context?(env, :struct_reference)
-
-    defines_struct? = Intelligence.defines_struct?(env.project, module.full_name)
+    if Env.in_context?(env, :struct_reference) do
+      complete_in_struct_reference(env, builder, module)
+    else
+      detail = builder.fallback(module.summary, module.name)
+      completion(env, builder, module.name, detail)
+    end
+  end
 
-    immediate_descendent_structs =
-      immediate_descendent_struct_modules(env.project, module.full_name)
+  defp complete_in_struct_reference(%Env{} = env, builder, module) do
+    ancestors_and_more =
+      env.project
+      |> immediate_descendent_struct_modules(module.full_name)
+      |> ancestors_and_more()
 
-    defines_struct_in_descendents? =
-      immediate_descendent_defines_struct?(env.project, module.full_name) and
-        length(immediate_descendent_structs) > 1
+    Enum.flat_map(ancestors_and_more, fn {child_module_name, more} ->
+      local_name = local_module_name(module.full_name, child_module_name, module.name)
 
-    cond do
-      struct_reference? and defines_struct_in_descendents? and defines_struct? ->
-        more = length(immediate_descendent_structs) - 1
+      env
+      |> Translations.Struct.completion(builder, local_name, child_module_name, more)
+      |> List.wrap()
+    end)
+  end
 
-        [
-          Translations.Struct.completion(env, builder, module.name, module.full_name, more),
-          Translations.Struct.completion(env, builder, module.name, module.full_name)
-        ]
+  defp ancestors_and_more(modules) do
+    modules
+    |> group_by_ancestor()
+    |> Enum.map(fn {ancestor, descendants} ->
+      {ancestor, length(descendants)}
+    end)
+  end
 
-      struct_reference? and defines_struct? ->
-        Translations.Struct.completion(env, builder, module.name, module.full_name)
+  defp group_by_ancestor(modules) do
+    sorted = Enum.sort(modules)
```

Comment:
I still do not understand this approach. It feels very, very complicated

Correct me if I'm wrong, but: Say you're completing `Lexical.|`.  Let's focus on the `Lexical.Document` completion. Don't you have the following  non-exclusive cases:
1. Lexical.Document contains a struct. 
2. Lexical.Document is the ancestor of some structs

If 1 is true, then include the completion for `%Lexical.Document{}` in the response. If it isn't, don't do anything.
If 2 is true, ask intelligence how many structs have `Lexical.Document` as an ancestor. This is as simple as:
```elixir
def descendant_struct_count(%Project{} =project,  module) do 
  project
  |> Intelligence.collect_struct_modules(module, to: :infinity)
  |> Enum.count()
end
```

I tested this approach with the following code, and it worked: `Intelligence.collect_struct_modules(project, Lexical.Document, to: 500)` which returned `["Lexical.Document.Lines", "Lexical.Document.Range", "Lexical.Document",
 "Lexical.Document.Location", "Lexical.Document.Changes",
 "Lexical.Document.Position", "Lexical.Document.Edit"]`. 


(this would mean adding handling of `:infinity` to Inteliigence, but it seems a vastly simpler and more reusable approach)

Then after you have the ancestor count, and it's > 0, add the ancestor completions.

Reply:
The problem is that they have ancestral relationships themselves, and we cannot guarantee where the first ancestor will appear. It could be in the first layer, like `ProjectStruct`, or it could be in the last layer like `Project.Structs.User`.

so all I want is a map like this:

```elixir
  @modules [
    "Lexical.Document.Lines",
    "Lexical.Document.Range",
    "Lexical.Document",
    "Lexical.Server.State",
    "Lexical.Server.Configuration",
    "Lexical.Document.Location",
    "Lexical.Document.Changes",
    "Lexical.Document.Position",
    "Lexical.Document.Edit",
    "Lexical.Project"
  ]

  @expect %{
    "Lexical.Document" => [
      "Lexical.Document.Range",
      "Lexical.Document.Position",
      "Lexical.Document.Location",
      "Lexical.Document.Lines",
      "Lexical.Document.Edit",
      "Lexical.Document.Changes"
    ],
    "Lexical.Project" => [],
    "Lexical.Server.Configuration" => [],
    "Lexical.Server.State" => []
  }
  
 assert group_by_ancestor(@modules) == @expect

```